### PR TITLE
Support transcoding audio formats when saving tarred datasets (FLAC, OPUS)

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

Supports saving audio into compressed formats when creating a tarred dataset.

**Collection**: Any speech collection

# Changelog 
- Add option `--force_codec [name]` to `scripts/speech_recognition/convert_to_tarred_audio_dataset.py` 
- Acceptable values: `wav`, `flac`, `opus`, possibly others supported by libsndfile
- When `--force_codec` is provided, and the audio file extension is different than codec's name, we'll read the audio and transcode it in memory, then save to tar; otherwise the existing code path is selected
- Change the default to save files as `opus`. Soundfile uses 28kbps for OPUS which is adequate for podcasts/audiobooks according to https://wiki.xiph.org/Opus_Recommended_Settings -- the result is ~17x smaller than FLAC and 26x smaller than WAVE (current default). 

# Usage

```bash
python scripts/speech_recognition/convert_to_tarred_audio_dataset.py --manifest_path nemo_manifest.json --num_shards 4 --force_codec opus --workers 4 --max_duration 30
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: scripts/speech_recognition/convert_to_tarred_audio_dataset.py [hunk 1]
```diff
@@ -179,6 +182,15 @@
     action='store_true',
     help="Do not write sharded manifests along with the aggregated manifest.",
 )
+parser.add_argument(
```
Comment (titu1994) [inline scripts/speech_recognition/convert_to_tarred_audio_dataset.py:185]: Default should be None here

File: scripts/speech_recognition/convert_to_tarred_audio_dataset.py [hunk 2]
```diff
@@ -569,6 +582,25 @@ def _read_manifest(self, manifest_path: str, config: ASRTarredDatasetConfig):
 
         return entries, total_duration, filtered_entries, filtered_duration
 
+    def _write_to_tar(self, tar, audio_filepath: str, squashed_filename: str) -> None:
+        if (codec := self.config.force_codec) is None or audio_filepath.endswith(f".{codec}"):
```
Comment (titu1994) [inline scripts/speech_recognition/convert_to_tarred_audio_dataset.py:586]: This is python 3.10 syntax right ? Let's try to keep it runnable on lower python versions. Min version is 3.10 but mostly asr continues to work on 3.8 or so

**Review by titu1994 — COMMENTED** [review]
Looks great, just wrong default on the argparse

File: scripts/speech_recognition/convert_to_tarred_audio_dataset.py (hunk 2 above)
Comment (pzelasko) [reply scripts/speech_recognition/convert_to_tarred_audio_dataset.py:586]: It's python 3.8, we should be fine 

https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions

**Review by titu1994 — APPROVED** [review]
Thanks !

File: scripts/speech_recognition/convert_to_tarred_audio_dataset.py (hunk 2 above)
Comment (titu1994) [reply scripts/speech_recognition/convert_to_tarred_audio_dataset.py:586]: Perfect

**Comment by titu1994** [comment]
jenkins

**Comment by pzelasko** [comment]
jenkins

**Pull request merged**